src/passes/StackIR.cpp

```diff
@@ -65,7 +69,39 @@ class StackIROptimizer {
     if (passOptions.optimizeLevel >= 3 || passOptions.shrinkLevel >= 1) {
       local2Stack();
     }
-    removeUnneededBlocks();
+    // Removing unneeded blocks is dangerous with GC, as if we do this:
+    //
+    //   (call
+    //     (rtt)
+    //     (block
+    //       (nop)
+    //       (i32)
+    //     )
+    //   )
+    // === remove inner block ==>
+    //   (call
+    //     (rtt)
+    //     (nop)
+    //     (i32)
+    //   )
+    //
+    // Then we end up with a nop that forces us to emit this during load:
+    //
+    //   (call
+    //     (block
+    //       (local.set
+    //         (rtt)
+    //       )
+    //       (nop)
+    //       (local.get)
+    //     )
+    //     (i32)
+    //   )
+    //
+    // However, that is not valid as an rtt cannot be set to a local.
+    if (!features.hasGC()) {
+      removeUnneededBlocks();
+    }
```

Comment:
Would it be feasible to fix this issue on a case by case basis by looking at what type would need to be assigned to the local?

Reply:
I think it's possible, but it's kind of annoying, since it's the type of the local "behind" us... that is, the type right before the block that we actually remove.